### ISSUE TITLE
Question about 3D modeling

### ISSUE THREAD

**Issue by yulongovo** [ar4/deepwave#104]
I'm sorry to disturb you. I would like to ask if you have any recommended 3D wave propagation module developed based on pytorch?

**Comment by yulongovo** [comment]
Thank you for your reply! This is a exciting new, and I’m willing to wait. The 2D wave propagation simulation module you developed provides an excellent user experience, and I’ve found it very helpful.

**Comment by yulongovo** [comment]
What a coincidence! The improved vacuum formulation  method proposed by Zeng et al. (2012) performs very well. Zeng is a senior colleague from my group. He has a profound understanding of forward modeling and waveform inversion. Previously, I implemented the free-surface condition in the presence of surface topography by Fortran.

**Comment by ar4** [comment]
That is indeed a coincidence. I have been evaluating these methods and others for quite a while. Although these two seem like the most suitable for Deepwave, I have been encountering problems with stability after very large numbers of time steps in some cases. I am starting a fresh implementation now to see if the problem persists. If it does, and if you have time, I would be grateful if I could describe my implementation to you (or perhaps provide a simple version) to verify that I have implemented it correctly. I know Zeng discussed using an MPML-based PML for stability in another paper, but I would prefer to stay with the CPML method if possible. If the instability cannot be resolved, I may add the option to apply sponge damping on top of the PML to prevent instabilities from developing.

**Comment by yulongovo** [comment]
Please forgive me for replying to the message late. Due to the time difference, I'm already sleeping, Ha ha.Your thorough grasp of forward modeling is truly impressive. I will support your work to the best of my ability. I haven’t delved into forward simulation or kept up with the latest developments for some time. At present I mainly apply the package you developed to assist my waveform inversion. Please forgive me if I’m behind on forward simulation. To my knowledge, the classical PML algorithm becomes unstable when the Poisson's ratio of the medium is high. The MPML proposed by Zeng effectively mitigates problems with stability after very large numbers of time steps in high-Poisson’s-ratio cases.

**Comment by ar4** [comment]
It is taking longer than I hoped to complete this. The propagators are almost finished, but I still need to do more testing and implement a method for compressing gradients and saving them to disk (which will be important for inversion with the 3D propagator). I apologise for the delay, and I hope that it is not causing you a lot of inconvenience.

**Comment by yulongovo** [comment]
Congratulations! I feel I may have caused you some trouble, and I really appreciate how seriously you’ve treated my questions and the help you’ve provided. I’ve already started using your updated version of deepwave, and everything seems to be working perfectly so far. I truly hope deepwave continues to gain recognition and attract more users. I’m also very excited about the upcoming 3D version! If there is anything I can do to help, please let me know.




一场空
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "ar4/deepwave"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年10月29日(星期三) 凌晨3:50
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [ar4/deepwave] Question about 3D modeling (Issue #104)



ar4 left a comment (ar4/deepwave#104)
 
It is taking longer than I hoped to complete this. The propagators are almost finished, but I still need to do more testing and implement a method for compressing gradients and saving them to disk (which will be important for inversion with the 3D propagator). I apologise for the delay, and I hope that it is not causing you a lot of inconvenience.

**Comment by Costaki33** [comment]
Hi Alan,

Thank you so much for creating DeepWave. My name is Constantinos Skevofilax and I'm a Computational Engineer at the Texas Seismological Network. We are really impressed with your work and starting to implement into our training flow for FWI models for tomographical imaging. 

I saw your comments with @yulongovo about implementing 1D and 3D propagation into DeepWave. Thank you for this, we're excited for this so we can begin using DeepWave in 3D. We are starting to implement it using our 2D solution. 

If we can be of assistance in any way, please let us know. 

Best, 
Constantinos Skevofilax
Texas Seismological Network

**Comment by ar4** [comment]
Hi Constantinos,

Thank you for your kind words. It means a lot to me to know that people find the package useful. Please let me know if you ever have any feedback. This could include things that don't behave the way you expect, suggestions for improvements (even as small as typos), and ideas for how to improve the code.

**Comment by Costaki33** [comment]
Thanks Alan for getting back to me and will do. 

I have a quick question, do you happen to have a timeline regarding when the 3D implementation will be released?

**Comment by Vcholerae1** [comment]
Hi Alan,

DeepWave is a really exciting tool – thanks for making it available.

I’ve heard that you’re working on gradient compression and disk I/O solutions for 3D wave propagation, and I’d love to get your thoughts. In particular, I’m thinking about trying two ideas in DeepWave:
– using random trace estimation ([arXiv:2104.00794](https://arxiv.org/abs/2104.00794)
), and
– using RWII ([Geo 2020-0606](https://library.seg.org/doi/10.1190/geo2020-0606.1)
).

Do you think these would be reasonably straightforward to implement in DeepWave, or are there important limitations in the current design that I should keep in mind?

I’d really appreciate any guidance you’re willing to share. Thanks a lot for your time.

Best,

**Comment by ar4** [comment]
I have finally released the new version (v0.0.24) with 1D, 2D, and 3D propagation, and greater control over storage of the intermediate data computed during forward propagation that is needed during backpropagation to calculate gradients with respect to the input property models. The intermediate data can now be compressed (currently using a simple and lossy 8 bits/element method), and can be stored on the compute device (as before), in CPU memory, or on disk. It is also possible to precent Deepwave from storing this intermediate data and calculating gradients with respect to property models, to avoid unnecessary memory usage and computation if you wish to handle these tasks yourself using the callbacks feature.

Please let me know if you have any feedback. I hope that there aren't any bugs, but there are a lot of changes in this release, so there is a risk that some slipped through. Please let me know if you find any.

I would like to apologise again for this release taking so long. I hope that it didn't cause you great inconvenience.

**Comment by Costaki33** [comment]
Hi Alan, congratulations on releasing the new versions! No worries at all about the inconvenience, and I think I can say for the community we are more than happy to provide feedback to you so we can help make the code better. 

Thanks for all that you do.
